src/Monarch/Document.ts

```diff
@@ -0,0 +1,161 @@
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, version 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import 'setimmediate'
+
+import { OutputHandlersList } from 'monarch/Monarch/VirtualDom/OutputHandlersList'
+import { VirtualDomTree } from 'monarch/Monarch/VirtualDom/VirtualDomTree'
+import { unsafe_uncurried_applyPatchTree } from 'monarch/Monarch/VirtualDom/PatchTree'
+import { unsafe_uncurried_mount } from 'monarch/Monarch/VirtualDom'
+import {
+    DiffWorkEnvironment,
+    DiffWork,
+    DiffWorkResult,
+    mkRootDiffWork,
+    unsafe_uncurried_performDiffWork,
+} from 'monarch/Monarch/VirtualDom/DiffWork'
+import { mkScheduler } from 'monarch/Monarch/Scheduler'
+
+interface DispatchMessage<message> {
+    (message: message): Effect<Unit>
+}
+
+interface DispatchOutput<output> {
+    (output: output): Effect<Unit>
+}
+
+// prettier-ignore
+interface Spec<input, model, message, output> {
+    input: input
+    init(input: input): model
+    update: (message: message) => (model: model) => model
+    view(model: model): VirtualDomTree<message>
+    container: HTMLElement
+    runCommand: (message: message) => (model: model) => (dispatchMessage: DispatchMessage<message>) => (dispatchOutput: DispatchOutput<output>) => Effect<Unit>
+    onOutput(output: output): Effect<Unit>
+}
+
+interface State<model, message> {
+    commitedVirtualDomTree: VirtualDomTree<message>
+    diffWork?: DiffWork<any, any>
+    diffWorkResult?: DiffWorkResult<message>
+    hasRequestedAsyncRendering: boolean
+    model: model
+}
+
+function unsafe_documentImpl<input, model, message, output>({
+    init,
+    input,
+    update,
+    runCommand,
+    container,
+    view,
+    onOutput,
+}: Spec<input, model, message, output>): void {
+    const initialModel: model = init(input)
+    const initialVirtualDomTree: VirtualDomTree<message> = view(initialModel)
+    const outputHandlers = OutputHandlersList.mkNil(unsafe_dispatchMessage)
+    const environment: DiffWorkEnvironment<message, message> = {
+        scheduler: mkScheduler(),
+        unsafe_dispatchDiffWork,
+        unsafe_finishDiffWork,
+    }
+
+    let state: State<model, message> = {
+        commitedVirtualDomTree: initialVirtualDomTree,
+        diffWork: undefined,
+        diffWorkResult: undefined,
+        hasRequestedAsyncRendering: false,
+        model: initialModel,
+    }
+
+    const dispatchMessage: DispatchMessage<message> = message => () => unsafe_dispatchMessage(message)
+
+    const dispatchOutput: DispatchOutput<output> = output => () => {
+        onOutput(output)()
+    }
+
+    function unsafe_dispatchMessage(message: message): void {
+        const previousModel = state.model
+        const nextModel = update(message)(previousModel)
+
+        if (previousModel === nextModel) return
+
+        runCommand(message)(nextModel)(dispatchMessage)(dispatchOutput)()
```

Comment:
This is a bug. Execution of the command is dependent on the `message`. On every message the command should run.